---
title: "Metastudy: activation of the mTOR signalling pathway in bone marrow-derived macrophages and microglia"
output: html_notebook
---

```{r setup}
require("knitr")
opts_knit$set(root.dir = normalizePath('..'))

library(edgeR)
library(DESeq2)
library(VennDiagram)
library(vsn)
library("dplyr")
library("ggplot2")

```

# Outline

* When the blood brain barrier is disrupted, bone marrow-derived macrophages (BMDM) can migrate across into the brain tissue.
* Disruption can be caused by a tumour or irradiation.
* They join the existing population of microglia (MG).
* The two cell types are difficult to distinguish.
* There is epigenetic evidence that certain pathways do differ between BMDM and MG.
* Both cell types seem to acquire a new 'education' in a tumour environment. They are known as tumour-associated (TA).

# Hypothesis
*The mTOR pathway is more active in TA-MG (relative to healthy MG) than in TA-BMDM (relative to healthy BMDM).*

# Available data

Two public mouse RNA-Seq datasets retrieved from GEO.

GSE68376: _Genetic Cell Ablation Reveals Clusters of Local Self-Renewing Microglia in the Mammalian Central Nervous System_ Bruttger, J. et al. Immunity (2015).  All samples from a healthy animal.  3 x BMDM biological replicates; 2 x WT MG biological replicates; 3 x repopulating MG biological replicates (following genetic ablation).

GSE86573: _Macrophage Ontogeny Underlies Differences in Tumor-Specific Education in Brain Malignancies_ Bowman et al. Cell Reports (2016). 3 x BMDM biological replicates and 3 x MG biological replicates from 2 different mouse models (total of 12 samples, 6 of each cell type).

# Aim

Use differential expression analysis of the 'gene count' data to test the hypothesis. Identify significantly DE genes in TA-BMDM vs healthy BMDM and TA-MG vs healthy MG. Then find genes that only appear in one or the other list. Finally, look for enrichment in mTOR pathway.

# Analysis

Import necessary libraries and load data. We set the metadata here too. *NB: the monocytes are labelled as BMDM - it that OK?*. We'll set the FDR and minimum effect size (absolute fold change) here for all comparisons. I've added a 'composite' column to the metadata, which will make it easier to define comparisons in the analysis.

```{r, message=FALSE, warning=FALSE, results="hide"}
fdr <- 0.05
logFC <- 1  # FC = 2

source('io/output.R')
source('_settings.R')

in.file <- file.path(data.dir.raid, "rnaseq", "GSE86573", "GSE86573_raw_counts.csv")
data.bowman <- read.csv(in.file, row.names = 1)

in.file <- file.path(data.dir.raid, "rnaseq", "GSE68376", "GSE86573_raw_counts.csv")
data.bruttger <- read.csv(in.file, row.names = 1)

##TODO

# we'll change the sample names for clarity
colnames(data) <- c("gl261_bmdm_rep1", "gl261_mg_rep1", "gl261_bmdm_rep2", "gl261_mg_rep2", "gl261_bmdm_rep3", "gl261_mg_rep3",
                    paste0("monocyte_rep", seq(1:5)), paste0("mg_rep", seq(1:3)),
                    "gemm_bmdm_rep1", "gemm_mg_rep1", "gemm_bmdm_rep2", "gemm_mg_rep2", "gemm_bmdm_rep3", "gemm_mg_rep3")
# set the group metadata
meta <- data.frame(row.names = colnames(data))
meta$cell_type = as.factor(c(
  rep(c("bmdm", "mg"), 3), rep("bmdm", 5), rep("mg", 3), rep(c("bmdm", "mg"), 3)
))
meta$model = as.factor(c(
  rep('gl261', 6), rep('gemm', 8), rep('gemm', 6)
))  # healthy cells come from a GEMM background?
meta$disease = as.factor(c(
  rep('tumour', 6), rep('healthy', 8), rep('tumour', 6)
))
meta$group <- 'h-mg'
meta$group[(meta$cell_type == 'bmdm') & (meta$disease == 'tumour')] = 'ta-bmdm'
meta$group[(meta$cell_type == 'mg') & (meta$disease == 'tumour')] = 'ta-mg'
meta$group[(meta$cell_type == 'bmdm') & (meta$disease == 'healthy')] = 'h-bmdm'
meta$group <- as.factor(meta$group)
meta$composite <- paste(as.vector(meta[,1]), as.vector(meta[, 2]), as.vector(meta[, 3]), sep='_')
meta
```

Normalise the data and look at the number of genes. 

```{r}
dgel <- DGEList(counts = data, samples = meta)
dgel <- calcNormFactors(dgel)
sprintf("Number of genes in raw data: %i", nrow(dgel))
```

Some of these have very low expression. Filtering them out now will help reduce the number of hypothesis tests we will carry out later, and increase the power accordingly. We should use counts per million (CPM) so that we take library size into account. In the smallest sample (~5mi reads), 1 CPM means 5 reads. In the largest sample (~30mi reads), 1 CPM means 30 reads. Let's require that at least 2 samples have CPM > 1 to retain a gene.

```{r}
keep <- rowSums(cpm(dgel) > 1) >= 2
dgel <- dgel[keep, ,  keep.lib.sizes=F]
sprintf("Number of genes in filtered data: %i. %.2f%% of genes retained.", nrow(dgel), sum(keep) / length(keep) * 100)
```

Let's look at the MDS plots. This can indicate batch effects and/or significant groupings.

```{r}
plotMDS(dgel)
title("MDS plot")
```

The first dimension separates monocytes from MG and BMDM. The second dimension is less clear: it separates MG and BMDM but very imperfectly.

We can also transform our data for the purpose of running other exploratory methods like PCA or hierarchical clustering. We could just use $\log_2(x+1)$, but there are other options including `rlog` and `vst`. Here, we take two samples and look at how their gene values compare under the three transformations.

```{r}

dds <- DESeqDataSetFromMatrix(countData = data, colData = meta, design = ~0 + group)
dds <- estimateSizeFactors(dds)


df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(rld)[, 1:2]) %>% mutate(transformation = "rlog"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))
  
colnames(df)[1:2] <- c("x", "y")  

ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation)  

```

Let's look at the PCA plot for `rlog` transformed data.

```{r}
rld <- rlog(dds, blind = FALSE)
plotPCA(rld, intgroup = c("cell_type", "disease", "model"))
```

All the groups cluster together. Principal component 1 is most important in splitting by cell type.  PC2 divides the two model types.

Now we can run `edgeR`. The most crucial part is defining the _design matrix_ and _contrasts_, which together encode the comparisons we want to make.

## `edgeR`: Which genes respond differently in TA-MG vs healthy MG? Which genes respond differently in TA-BMDM vs healthy monocytes?

Here we only include the combined terms: TA-MG, TA-BMDM, H-MG, H-BMDM.

```{r}
design <- model.matrix(~0 + cell_type:disease, data = meta)
dgel <- estimateDisp(dgel, design)
fit <- glmFit(dgel, design)
lrt <- glmTreat(fit, contrast=c(0, -1, 0, 1), lfc=logFC)
ta_mg.vs.h_mg <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-MG vs healthy MG. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_mg.vs.h_mg)[1], 
  fdr,
  sum(ta_mg.vs.h_mg$table$logFC < 0),
  sum(ta_mg.vs.h_mg$table$logFC > 0)
)

lrt <- glmTreat(fit, contrast=c(-1, 0, 1, 0), lfc=logFC)
ta_bmdm.vs.h_bmdm <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-BMDM vs healthy monocytes. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_bmdm.vs.h_bmdm)[1], 
  fdr,
  sum(ta_bmdm.vs.h_bmdm$table$logFC < 0),
  sum(ta_bmdm.vs.h_bmdm$table$logFC > 0)
)
```
Look for the intersection.
```{r}
area.1 <- nrow(ta_mg.vs.h_mg$table)
area.2 <- nrow(ta_bmdm.vs.h_bmdm$table)
area.cross <- length(intersect(rownames(ta_bmdm.vs.h_bmdm), rownames(ta_mg.vs.h_mg)))
grid.newpage()
venn <- draw.pairwise.venn(area.1, area.2, area.cross, category = c("TA-MG vs healthy MG", "TA-BMDM vs healthy monocytes"),
                             lty = rep("blank", 2), fill = c("light blue", "pink"), alpha = rep(0.5, 2)
)
```

For my own records: this is _exactly_ the same as using group variable and carrying out the fitting on that instead.

```{r}
design <- model.matrix(~0 + group, data = meta)
colnames(design) <- sub('group', '', colnames(design))
dgel <- estimateDisp(dgel, design)
fit <- glmFit(dgel, design)
lrt <- glmTreat(fit, contrast=c(0, -1, 0, 1), lfc=logFC)
ta_mg.vs.h_mg <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-MG vs healthy MG. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_mg.vs.h_mg)[1], 
  fdr,
  sum(ta_mg.vs.h_mg$table$logFC < 0),
  sum(ta_mg.vs.h_mg$table$logFC > 0)
)

lrt <- glmTreat(fit, contrast=c(-1, 0, 1, 0), lfc=logFC)
ta_bmdm.vs.h_bmdm <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-BMDM vs healthy monocytes. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_bmdm.vs.h_bmdm)[1], 
  fdr,
  sum(ta_bmdm.vs.h_bmdm$table$logFC < 0),
  sum(ta_bmdm.vs.h_bmdm$table$logFC > 0)
)
```

That's a much larger number than Bowman et al. reported. Why? They used DESeq2. Let's try that.

## `DESeq2`: Which genes respond differently in TA-MG vs healthy MG? Which genes respond differently in TA-BMDM vs healthy monocytes?

Start similarly, filtering out genes with low counts in most samples.

```{r}
samples <- colnames(data)
dds <- DESeqDataSetFromMatrix(countData = data[, samples], colData = meta[samples,], design = ~composite)
keep <- rowSums(counts(dds) > 5) >= 2
dds <- dds[keep,]
sprintf(
  "Number of genes in raw data: %i. Number after filtering: %i. %.02f %% retained.",
  length(keep),
  sum(keep),
  sum(keep) / length(keep) * 100.
)
dds <- estimateSizeFactors(dds)
des <- DESeq(dds, parallel = T)
for (t in c("mg", "bmtm")){
  res <- results(des, contrast=list(paste0("groupta.", t), paste0("grouph.", t)), cooksCutoff = F)
  padj <- res$padj
  padj[is.na(padj)] = 1.
  res <- res[(padj < fdr) & (abs(res$log2FoldChange) > logFC),]
}


res.mg <- results(des, contrast=list("groupta.mg", "grouph.mg"))
res.bmdm <- results(des, contrast=list("groupta.bmdm", "grouph.bmdm"))

```




```{r}
meta$cell_type <- relevel(meta$cell_type, ref='mg')
meta$disease <- relevel(meta$disease, ref='healthy')
design <- model.matrix(~disease * cell_type, data = meta)

dgel <- estimateDisp(dgel, design)
fit <- glmFit(dgel, design)
lrt <- glmTreat(fit, contrast=c(0, -1, 0, 1), lfc=logFC)
ta_mg.vs.h_mg <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-MG vs healthy MG. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_mg.vs.h_mg)[1], 
  fdr,
  sum(ta_mg.vs.h_mg$table$logFC < 0),
  sum(ta_mg.vs.h_mg$table$logFC > 0)
)

lrt <- glmTreat(fit, coef = 2, lfc=logFC)
ta_bmdm.vs.h_bmdm <- topTags(lrt, n=Inf, p.value=fdr)  # p value here is the FDR
sprintf(
  "TA-BMDM vs healthy monocytes. Found %i DE genes at FDR=%.03f. Of these, %i are down and %i are up", 
  dim(ta_bmdm.vs.h_bmdm)[1], 
  fdr,
  sum(ta_bmdm.vs.h_bmdm$table$logFC < 0),
  sum(ta_bmdm.vs.h_bmdm$table$logFC > 0)
)
```


Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).
