---
title: "Implementing multiple Differential Methylation Analysis methods"
author: "Gabriel Rosser"
date: "20 February 2019"
output: html_document
---

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = TRUE)
opts_knit$set(root.dir = normalizePath('..'))

knitr::opts_chunk$set(echo = TRUE)

require(limma)
require("ChAMP")
require("minfi")
require("wateRmelon")
require("data.table")
require(sva)
require(RColorBrewer)
require(openxlsx)
require(ks)
require("ggplot2")
require("gridExtra")
require("reshape")
require("doParallel")

# this specifies the normalisation algorithm to use
norm.fun <- 'swan'
# norm.fun <- 'bmiq'

# this is our FDR cutoff
alpha <- 0.01

# This specifies a minimum logFC
min_lfc <- 1.

# see description below
# load_preproc <- F
load_preproc <- T

# initialise multiple cores
registerDoParallel(cores = 4)


```

## Some useful functions

Here I've defined useful (and reusable) functions for later in the script.

```{r}
bar_plot <- function(dmr_res, fc_var = 'logFC', pids=NULL, ttl = NULL) {
  if (is.null(pids)) pids <- names(dmr_res)

  df <- data.frame(row.names = c('Hyper', 'Hypo'))
  for (p in pids) {
    n_pos <- sum(dmr_res[[p]][, fc_var] > 0)
    n_neg <- sum(dmr_res[[p]][, fc_var] < 0)
    df[, p] <- c(n_pos, n_neg)
  }
  df.pct <- t(t(df) / colSums(df)) * 100.
  
  df.long <- melt(as.matrix(df))
  colnames(df.long) <- c('Direction', 'Patient', 'Value')
  pid_col <- sprintf("%03d", df.long$Patient)
  df.long$Patient <- factor(pid_col, levels = pids)
  
  df.pct.long <- melt(as.matrix(df.pct))
  colnames(df.pct.long) <- c('Direction', 'Patient', 'Value')
  df.pct.long$Patient <- factor(pid_col, levels = pids)
  
  # numbers chart
  plt2 <- ggplot(df.long, aes(x=Patient, y=Value, fill=Direction)) + geom_bar(stat='identity') + scale_fill_manual(values=c("#FF381F", "#89CD61")) + labs(y="Number")
  
  # %
  plt1 <- ggplot(df.pct.long, aes(x=Patient, y=Value, fill=Direction)) + geom_bar(stat='identity') + scale_fill_manual(values=c("#FF381F", "#89CD61")) + labs(title=ttl, y='% by direction')
  
  plt.main <- grid.arrange(plt1, plt2, nrow=2)

}

get_specific_results <- function(dmr_res, pids = NULL, ident_fun = rownames) {
  if (is.null(pids)) pids <- names(dmr_res)
  pnames <- lapply(dmr_res, ident_fun)
  specific <- lapply(pids, function(p) {setdiff(pnames[[p]], Reduce(union, pnames[pids != p]))})
  names(specific) <- pids

  res_spec <- list()
  for (p in pids) {
    res_spec[[p]] <- dmr_res[[p]][pnames[[p]] %in% specific[[p]],]
  }
  return(res_spec)
}
```

## Load data and prepare metadata

Two options here:

1. Load the data from the raw `idat` files. This ensures that all samples are normalised together for each preparation type. However, it doesn't reflect the DMR process I have used in the past [`load_preproc = FALSE`].
2. Load the data from pre-processed `csv` files. Data were normalised in the batches as they arrived [`load_preproc = TRUE`]. This is quite a lot faster.

**NB** I have tested the effect of these two approaches on the downstream analysis in Python and the results are **not** affected. Therefore, option (2) seems easier.

```{r echo=FALSE}
source('_settings.R')
source("methylation/loader.R")

pids <- c('018', '019', '030', '031', '017', '050', '054', '061', '026', '052')

cc_samples <- c(
  "GBM018_P12",
  "GBM018_P10",
  "DURA018_NSC_N4_P4",
  "DURA018_NSC_N2_P6",
  "GBM019_P4",
  "GBM019_P3n6",
  "DURA019_NSC_N8C_P2",
  "DURA019_NSC_N5C1_P2",
  "GBM030_P9",
  "GBM030_P5",
  "DURA030_NSC_N16B6_P1",
  "DURA030_NSC_N9_P2",
  "GBM031_P7",
  "GBM031_P4",
  "DURA031_NSC_N44B_P2",
  "DURA031_NSC_N44F_P3",
  "GBM017_P3",
  "GBM017_P4",
  "DURA017_NSC_N3C5_P4",
  "GBM050_P7n8",
  "GBM050_P9",
  "DURA050_NSC_N12_P3",
  "DURA050_NSC_N16_P4",
  "GBM054_P4",
  "GBM054_P6",
  "DURA054_NSC_N3C_P2",
  "DURA054_NSC_N2E_P1",
  "GBM061_P3",
  "GBM061_P5",
  "DURA061_NSC_N4_P2",
  "DURA061_NSC_N1_P3n4",
  "GBM026_P8",
  "GBM026_P3n4",
  "DURA026_NSC_N31D_P5",
  "GBM052_P6n7",
  "GBM052_P4n5",
  "DURA052_NSC_N4_P3",
  "DURA052_NSC_N5_P2"
)

cc_base.dirs <- c(
  file.path(data.dir.raid, 'methylation', "2017-05-12"),
  file.path(data.dir.raid, 'methylation', "2016-12-19_ucl_genomics"),
  file.path(data.dir.raid, 'methylation', "2017-09-19"),
  file.path(data.dir.raid, 'methylation', "2018-01-12"),
  file.path(data.dir.raid, 'methylation', "2017-08-23"),
  file.path(data.dir.raid, 'methylation', "2018-04-09")
)

if (load_preproc) {
  cc_beta <- list()
  snames <- NULL
  batches <- NULL
  for (c in cc_base.dirs) {
    this_batch <- split_path(c)[1]
    
    the_fname <- file.path(c, 'beta', paste0('beta_', norm.fun, '.csv.gz'))
    if (! file.exists(the_fname)) {
      stop(sprintf("File not found: %s", the_fname))
    }
    # get header
    hdr <- scan(the_fname, sep=',', nlines=1, what = 'character')
    ix <- hdr %in% cc_samples
    load_cols <- c(1, which(ix))
    snames <- c(snames, hdr[ix])
    batches <- c(batches, rep(this_batch, sum(ix)))
    
    # load only selected data
    this_dat <- fread(sprintf("zcat %s", the_fname), select = load_cols, data.table = F)
    
    # set row names and drop the unneeded col
    rownames(this_dat) <- this_dat[,1]
    this_dat$V1 <- NULL
    
    cc_beta[[c]] <- this_dat
  }
  # reduce to common probes
  common_probes <- Reduce(intersect, lapply(cc_beta, rownames))
  cc_beta <- lapply(cc_beta, function(x) {as.data.frame(t(x[common_probes,]))})
  # bind by row then transpose and restore colnames
  cc_beta <- as.data.frame(t(rbindlist(cc_beta)))
  colnames(cc_beta) <- snames

} else {
  
  lst <- get_filenames_from_batches(cc_base.dirs, cc_samples)

  cc_res <- process_idats(lst$in.files, lst$snames, norm.fun=norm.fun, force = T)
  cc_beta <- cc_res$beta
  
  snames <- lst$snames
  batches <- lst$batches
  
}

cc_m <- MfromBeta(cc_beta)

cc_meta <- data.frame(row.names = snames)
cc_meta$batch <- batches
cc_meta$patient_id <- gsub("[^0-9]*([0-9]{3}).*", "\\1", rownames(cc_meta), perl=T)
cc_meta$cell_type <- "iNSC"
cc_meta[grep('GBM', rownames(cc_meta)), 'cell_type'] <- 'GIC'

if (!load_preproc) {
  # save the data to a file, which we can use from within Python
  write.csv(cc_meta,"hgic_gbm_insc_meta.csv")
  write.csv(cc_m, sprintf("hgic_gbm_insc_mvals_%s.csv", norm.fun))
}

# annotation
RSobject <- RatioSet(cc_beta, annotation = c(array = "IlluminaHumanMethylationEPIC",annotation = "ilm10b4.hg19"))
anno <- getAnnotation(RSobject)

```
## `limma` {limma}

This package allows us to build a GLM and identify differentially methylated probes (DMPs). This package was originally designed for gene expression microarrays, but it is also applicable to methylation arrays. We're going to mimic the DE approach by pooling all samples for the purpose of dispersion estimation, then iterating over patient-specific contrasts.

```{r limma}
include_batch_var <- T
pid <- cc_meta$patient_id
ct <- cc_meta$cell_type
group <- paste(pid, ct, sep = '_')

if (include_batch_var) {
  bs <- gsub("-", "_", batches)
  design <- model.matrix(~0 + group + bs)
} else {
  design <- model.matrix(~0 + group)
}

contrast_string = c()
for (p in pids) {
  a <- paste0("group", p, "_GIC")
  b <- paste0("group", p, "_iNSC")
  contrast_string <- c(contrast_string, paste(a, b, sep = '-'))
}
the_contrasts <- makeContrasts(contrasts = contrast_string, levels=design)

fit.reduced <- lmFit(cc_m, design)

fit.reduced <- contrasts.fit(fit.reduced, the_contrasts)
fit.reduced <- eBayes(fit.reduced, trend = T, robust=T)

res_limma <- list()
for (i in seq(ncol(the_contrasts))) {
  p <- pids[i]
  res_limma[[p]] <- topTable(fit.reduced, coef=i, n=Inf, lfc=min_lfc, p.value = alpha)
}
```

Look at the number (or proportion) of DMPs by direction.

```{r fig.width=5, fig.height=8}

# full list
plt <- bar_plot(res_limma)
ggsave("full_dmps_limma.png", plot = plt, device = 'png')

# specific list
res_limma.spec <- get_specific_results(res_limma)

plt <- bar_plot(res_limma.spec)
ggsave("specific_dmps_limma.png", plot = plt, device = 'png')

```

## `limma` without pooling {limma_nopool}

What if we _don't_ pool samples, but re-run the comparison separately for each patient?

```{r}
res_limma.sep <- list()

for (p in pids) {
  ix <- cc_meta$patient_id == p
  this_dat <- cc_m[, ix]
  this_meta <- cc_meta[ix, ]
  ct <- this_meta$cell_type
  design <- model.matrix(~0 + ct)
  
  fit.reduced <- lmFit(this_dat, design)
  fit.reduced <- contrasts.fit(fit.reduced, makeContrasts('ctGIC-ctiNSC', levels = design))
  fit.reduced <- eBayes(fit.reduced, trend = T, robust=T)
  
  res_limma.sep[[p]] <- topTable(fit.reduced, coef=1, n=Inf, lfc=min_lfc, p.value = alpha)
}

```

There are *no* DMPs for patient 026, so this is not a good approach. Look at the numbers / directions:

```{r fig.width=5, fig.height=8}
# full list
plt <- bar_plot(res_limma.sep, pids = pids[pids != '026'])
ggsave("full_dmps_limma_separate.png", plot = plt, device = 'png')

# specific list
res_limma.sep.spec <- get_specific_results(res_limma.sep)

plt <- bar_plot(res_limma.sep.spec, pids = pids[pids != '026'])
ggsave("specific_dmps_limma_separate.png", plot = plt, device = 'png')

```

As we saw in the DE case, running the analysis without pooled dispersion results in very variable numbers of differential methylated probes.

## `DMRcate`

This workflow essentially proceeds by first running DMP analysis (in `limma`), then aggregating the outputs.

```{r, dmrcate, echo=F}
min_probes <- 6
min_delta_beta <- 0.3

if (include_batch_var) {
  bs <- gsub("-", "_", batches)
  design <- model.matrix(~0 + group + bs)
} else {
  design <- model.matrix(~0 + group)
}

contrasts <- do.call(makeContrasts, c(as.list(contrast_string), list(levels=design)))
colnames(contrasts) <- pids
res_dmrcate <- list()

for (p in pids) {
  this_annot <- cpg.annotate(
    "array",
    as.matrix(cc_m),
    what='M',
    arraytype = 'EPIC',
    analysis.type = 'differential',
    design = design,
    contrasts = T,
    cont.matrix = contrasts,
    coef = p
  )
  this_dmrcate <- dmrcate(this_annot, lambda = 1000, C=2, mc.cores = 8, betacutoff = min_delta_beta, min.cpgs = min_probes)
  res_dmrcate[[p]] <- this_dmrcate$results
}

```

Generate the same plots

```{r dmrcate_plots, fig.width=5, fig.height=8}
plt <- bar_plot(res_dmrcate, fc_var = 'meanbetafc')
ggsave("full_dmrs_dmrcate.png", plot = plt, device = 'png')

# specific list
res_dmrcate.spec <- get_specific_results(res_dmrcate, ident_fun = function(x) x[, 'coord'])

plt <- bar_plot(res_dmrcate.spec, fc_var = 'meanbetafc')
ggsave("specific_dmrs_dmrcate.png", plot = plt, device = 'png')

```

## Bumphunter

The `bumphunter` package is conveniently implemented in the `minfi` package, so we'll use that. The issue here is with the implementation: there is no obvious way to test specific contrasts (as I have done above). We need to specify our comparison as a single coefficient, which rules out including all data at once. We'll basically be running the [limma without pooling option](#limma_nopool) process. Still, let's give it a try?

```{r bumphunter, echo = F}
min_probes <- 6
max_gap <- 400
delta_m_cutoff <- 2.  # this is larger than I would use, but needs must: bumphunter is slow
nperm <- 1000  # the number of perms used by bumphunter
run_init_check <- F  # first time around, we should run a scan with no permutations to check that the number of candidates isn't outrageous

# since we'll be running bumphunter a number of times, it's best to define the regions once in advance
clusters <- clusterMaker(anno$chr, anno$pos, maxGap = max_gap)

# there are many clusters with few probes. Let's remove them?
tbl <- table(clusters)
too_small <- names(tbl)[tbl < min_probes]  # cluster numbers that need to be removed

# can't leave NA in clusters, as it results in an error in the bumphunter() call
# clusters[clusters %in% too_small] <- -1

ix_keep <- ! clusters %in% too_small
probes_keep <- rownames(anno)[ix_keep]

this_m <- cc_m[probes_keep,]
this_anno <- anno[probes_keep,]

# redefine clusters, all of which must now contain at least min_probes probes
clusters <- clusterMaker(this_anno$chr, this_anno$pos, maxGap = max_gap)

if (run_init_check) {
  p <- pids[1]
  ix <- cc_meta$patient_id == p
  this_dat <- this_m[, ix]
  this_meta <- cc_meta[ix, ]
  ct <- factor(this_meta$cell_type, levels=c('iNSC', 'GIC'))
  design <- model.matrix(~ct)
  
  this_initscan <- bumphunter(
    as.matrix(this_dat),
    design,
    chr = this_anno$chr,
    pos = this_anno$pos,
    coef = 2,
    cutoff = delta_m_cutoff,
    B = 0,
    type = 'M',
    cluster = clusters,
    maxGap = max_gap
  )
  # TODO: report result
  
}

res_bumphunter <- list()

for (p in pids) {
  ix <- cc_meta$patient_id == p
  this_dat <- this_m[, ix]
  this_meta <- cc_meta[ix, ]
  ct <- factor(this_meta$cell_type, levels=c('iNSC', 'GIC'))
  design <- model.matrix(~ct)
  
  this_res <- bumphunter(
    as.matrix(this_dat),
    design,
    chr = this_anno$chr,
    pos = this_anno$pos,
    coef = 2,
    cutoff = delta_m_cutoff,
    B = nperm,
    type = 'M',
    cluster = clusters,
    maxGap = max_gap
  )
  res_bumphunter[[p]] <- this_res$table
 
}

```



